src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,907 @@
+                              (name (buff 16))
+                              (salt (buff 20))
+                              (zonefile-hash (buff 20)))
+  (let (
+    (can-sender-register-name (try! (can-register-name tx-sender)))
+    (hashed-salted-fqn (hash160 (concat (concat (concat name 0x2e) namespace) salt))))
+    (let ( 
+        (preorder (unwrap!
+          (map-get? name-preorders ((hashed-salted-fqn hashed-salted-fqn) (buyer tx-sender)))
+          (err ERR_NAME_PREORDER_NOT_FOUND)))
+        (namespace-props (unwrap!
+          (map-get? namespaces ((namespace namespace)))
+          (err ERR_NAMESPACE_NOT_FOUND)))
+        (current-owner (nft-get-owner? names (tuple (name name) (namespace namespace)))))
+      ;; The name must only have valid chars
+      (asserts!
+        (not (has-invalid-chars name))
+        (err ERR_NAME_CHARSET_INVALID))
+      ;; The name must not exist yet, or be expired
+      (if (is-none current-owner)
+        true
+        (asserts!
+          (unwrap! (is-name-lease-expired? namespace name) (err ERR_PANIC))
+          (err ERR_NAME_UNAVAILABLE)))
+      ;; The name's namespace must be launched
+      (asserts!
+        (is-some (get launched-at namespace-props))
+        (err ERR_NAMESPACE_NOT_LAUNCHED))
+      ;; The preorder must have been created after the launch of the namespace
+      (asserts!
+        (> (get created-at preorder) (unwrap! (get launched-at namespace-props) (err ERR_PANIC)))
+        (err ERR_NAME_PREORDERED_BEFORE_NAMESPACE_LAUNCH))
+      ;; The preorder entry must be unclaimed
+      (asserts!
+        (is-eq (get claimed preorder) false)
+        (err ERR_NAME_ALREADY_CLAIMED))
+      ;; Less than 24 hours must have passed since the name was preordered
+      (asserts!
+        (< block-height (+ (get created-at preorder) NAME_PREORDER_CLAIMABILITY_TTL))
+        (err ERR_NAME_CLAIMABILITY_EXPIRED))
+      ;; The amount burnt must be equal to or greater than the cost of the name
+      (asserts!
+        (>= (get stx-burned preorder) (compute-name-price name (get price-function namespace-props)))
+        (err ERR_NAME_STX_BURNT_INSUFFICIENT))
+      ;; The principal can register a name
+      (asserts!
+        can-sender-register-name
+        (err ERR_PRINCIPAL_ALREADY_ASSOCIATED))
+      ;; Mint the name if new, transfer the name otherwise.
+      (if (is-none current-owner)
+        (begin
+          (unwrap! 
+            (nft-mint? 
+              names 
+              (tuple (namespace namespace) (name name)) 
+              tx-sender)
+            (err ERR_NAME_COULD_NOT_BE_MINTED))
+          (map-set rule-1-1
+            ((owner tx-sender))
+            ((namespace namespace) (name name))))
+        (if (is-eq tx-sender (unwrap! current-owner (err ERR_PANIC)))
+          true
+          (let ((previous-owner (unwrap! current-owner (err ERR_PANIC)))) 
+            (unwrap!
+              (update-name-ownership? namespace name previous-owner tx-sender)
```

Comment:
@lgalabru This will always fail because nft-transfer fails with err u1 as the tx-sender is not hte owner of the nft, won't it?

To me it looks like it is only possible to re-register an expired name if the previous owner actively gives it back. Is that right?

Reply:
The transfer is being performed from the `previous-owner` to the `tx-sender`.

Reply:
@lgalabru `nft-transfer?` will fail with `u1` if the token is not owned by the sender.

Reply:
@friedger It will work -- `nft-transfer?` will succeed even if `previous-owner` is not `tx-sender`.  `nft-transfer?` will fail with `u1` only if `previous-owner` doesn't actually own the asset, regardless of whether or not `previous-owner` is equal to `tx-sender`.  You can see it here in the implementation: https://github.com/blockstack/stacks-blockchain/blob/1ad60a216e53473d6fb103e2e171cfe1003b9883/src/vm/functions/assets.rs#L348 (note that the method does not consider `env.sender`, which is the value of `tx-sender` in the current context).

This is also true for `ft-transfer?`, btw -- the sender does not need to be `tx-sender`.  The sender only needs to be `tx-sender` for `stx-transfer?` and `stx-burn?`.